sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala

```diff
@@ -150,7 +163,7 @@ object FileFormatWriter extends Logging {
     )
 
     // We should first sort by partition columns, then bucket id, and finally sorting columns.
-    val requiredOrdering = partitionColumns ++ bucketIdExpression ++ sortColumns
+    val requiredOrdering = partitionColumns ++ bucketSortExpression ++ sortColumns
```

Comment:
So when the `bucketSortExpression` is None. We should still compare `actualOrdering` with the original `requiredOrdering` which contains bucket id expression. Because now when any of them matches `actualOrdering`, we don't need to sort again.